File: utils/background_worker.py
import time
import datetime
from utils.state_manager import load_lkv, save_lkv
from utils.weather_connector import get_live_weather_by_station
from utils.macro_connector import GlobalMacroEngine
from agents.data_agent import GlobalIssueTracker, DataAgent
import pandas as pd
from db import get_db_connection

def get_active_countries():
    """
    DB(regions 테이블)에 등록된 지역 코드 정보를 기준으로 활성화된 국가 목록을 동적으로 반환합니다.
    - 예: KR-11 -> South Korea
    - 등록된 국가가 없는 경우 최소 South Korea 1개를 보장합니다.
    """
    try:
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute("SELECT DISTINCT region_code FROM regions")
        codes = [row["region_code"] for row in cursor.fetchall()]
        conn.close()
    except Exception as e:
        print(f"⚠️ [Background Worker] DB 조회 실패 (South Korea 기본값 사용): {e}")
        return ["South Korea"]

    # ISO 3166-2 주(Province/State) 코드 또는 지명 매핑
    countries = set()
    for code in codes:
        code_upper = str(code).upper()
        if code_upper.startswith("KR-"):
            countries.add("South Korea")
        elif code_upper.startswith("US-"):
            countries.add("United States")
        elif code_upper.startswith("CN-"):
            countries.add("China")
        elif code_upper.startswith("JP-"):
            countries.add("Japan")
        elif code_upper.startswith("GB-"):
            countries.add("United Kingdom")
        # 추가적인 국가 코드가 들어오는 경우 매핑 확장 가능

    return list(countries) if countries else ["South Korea"]


def run_worker():
    print("🚀 [Background Worker] 가동 시작...")
    macro_engine = GlobalMacroEngine()
    issue_tracker = GlobalIssueTracker()
    
    # DataAgent 초기화 (Trends 수집용)
    try:
        data_agent = DataAgent()
    except Exception as e:
        print(f"⚠️ [DataAgent 초기화 실패]: {e}")
        data_agent = None

    # 기상 관측소 매핑 (간소화)
    station_map = {
        "South Korea": {"id": "47159", "name": "Busan", "lat": 35.1017, "lon": 129.03},
        "United States": {"id": "72503", "name": "New York", "lat": 40.7128, "lon": -74.0060},
        "China": {"id": "54511", "name": "Beijing", "lat": 39.9042, "lon": 116.4074},
        "Japan": {"id": "47662", "name": "Tokyo", "lat": 35.6895, "lon": 139.6917},
        "United Kingdom": {"id": "03772", "name": "London", "lat": 51.5074, "lon": -0.1278},
    }

    while True:
        print(f"\n🕒 [{datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] 데이터 수집 사이클 시작...")
        state = load_lkv()
        
        for country in get_active_countries():
            print(f"Fetching data for {country}...")
            country_data = state.get(country, {})
            
            # 1. 날씨 수집
            if country in station_map:
                s_info = station_map[country]
                try:
                    weather = get_live_weather_by_station(s_info["id"], s_info["lat"], s_info["lon"])
                    if weather:
                        country_data["weather"] = weather
                except Exception as e:
                    print(f"  ❌ Weather failed for {country}: {e}")
            
            # 2. 매크로 데이터 수집
            try:
                macro_vector = macro_engine.fetch_unified_macro_vector(country)
                if macro_vector:
                    country_data["macro"] = macro_vector
            except Exception as e:
                print(f"  ❌ Macro failed for {country}: {e}")
                
            # 3. GDELT 수집
            try:
                gdelt_info = issue_tracker.fetch_supply_chain_risk_tone(target_country=country)
                if gdelt_info:
                    country_data["gdelt"] = gdelt_info
            except Exception as e:
                print(f"  ❌ GDELT failed for {country}: {e}")
                
            # 4. Google Trends 수집
            if data_agent:
                try:
                    # DataAgent의 _fetch_trend_signal은 내부적으로 키워드를 검색함
                    trends_info = data_agent._fetch_trend_signal()
                    if trends_info:
                        country_data["trends"] = trends_info
                except Exception as e:
                    print(f"  ❌ Trends failed for {country}: {e}")
            
            # 타임스탬프 갱신
            country_data["timestamp"] = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
            state[country] = country_data
            
            # 과도한 API 요청 방지를 위한 딜레이
            time.sleep(1)
            
        # 상태 저장
        save_lkv(state)
        print(f"💾 수집 완료 및 저장됨. 30분 대기...")
        time.sleep(1800)  # 30분 주기 (1800초)

if __name__ == "__main__":
    run_worker()

File: scm-agent-backend/src/main/java/com/sigma/scm/service/BackgroundWorker.java
package com.sigma.scm.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.sigma.scm.domain.LkvState;
import com.sigma.scm.domain.LkvStateId;
import com.sigma.scm.repository.LkvStateRepository;
import com.sigma.scm.repository.RegionRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import net.javacrumbs.shedlock.spring.annotation.SchedulerLock;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

@Service
@RequiredArgsConstructor
@Slf4j
public class BackgroundWorker {

    private final LkvStateRepository lkvStateRepository;
    private final RegionRepository regionRepository;
    private final BatchAnalysisProxyService batchAnalysisProxyService;
    private final ObjectMapper objectMapper = new ObjectMapper();

    // 기상 관측소 매핑 테이블
    private static final Map<String, Map<String, Object>> STATION_MAP = new HashMap<>();

    static {
        STATION_MAP.put("South Korea", Map.of("id", "47159", "name", "Busan", "lat", 35.1017, "lon", 129.03));
        STATION_MAP.put("United States", Map.of("id", "72503", "name", "New York", "lat", 40.7128, "lon", -74.0060));
        STATION_MAP.put("China", Map.of("id", "54511", "name", "Beijing", "lat", 39.9042, "lon", 116.4074));
        STATION_MAP.put("Japan", Map.of("id", "47662", "name", "Tokyo", "lat", 35.6895, "lon", 139.6917));
        STATION_MAP.put("United Kingdom", Map.of("id", "03772", "name", "London", "lat", 51.5074, "lon", -0.1278));
    }

    private List<String> getActiveCountries() {
        // regions 테이블의 region_code 접두사를 매핑해 활성 국가 판별
        List<String> countries = new ArrayList<>();
        regionRepository.findAll().forEach(region -> {
            String code = region.getRegionCode().toUpperCase();
            if (code.startsWith("KR-")) countries.add("South Korea");
            else if (code.startsWith("US-")) countries.add("United States");
            else if (code.startsWith("CN-")) countries.add("China");
            else if (code.startsWith("JP-")) countries.add("Japan");
            else if (code.startsWith("GB-")) countries.add("United Kingdom");
        });
        
        Set<String> uniqueCountries = new HashSet<>(countries);
        if (uniqueCountries.isEmpty()) {
            uniqueCountries.add("South Korea"); // 최소 1개 기본값 보장
        }
        return new ArrayList<>(uniqueCountries);
    }

    /**
     * 주기적 30분 수집 태스크 (ShedLock 적용)
     */
    @Scheduled(fixedDelay = 1800000)
    @SchedulerLock(name = "background_scraping_task", lockAtMostFor = "15m", lockAtLeastFor = "5m")
    public void runScrapingCycle() {
        log.info("Starting background scraping cycle...");
        executeCycle("SYSTEM");
    }

    /**
     * Pod 시작 시 크래시 복구 스캐너 (ShedLock을 이용하여 1회성 안전 초기화)
     */
    @Scheduled(initialDelay = 0, fixedDelay = Long.MAX_VALUE)
    @SchedulerLock(name = "crash_recovery_scanner", lockAtMostFor = "5m", lockAtLeastFor = "30s")
    public void runCrashRecovery() {
        log.info("Pod started. Executing initial Crash Recovery scan...");
        executeCycle("CRASH_RECOVERY_SCANNER");
    }

    private void executeCycle(String actor) {
        String companyId = "COMPANY_SIGMA"; // 기본 회사 코드
        List<String> activeCountries = getActiveCountries();

        for (String country : activeCountries) {
            log.info("[{}] Fetching SCM variables for country: {}", actor, country);
            
            Map<String, Object> countryData = new LinkedHashMap<>();

            // 1. 날씨 정보 수집 (관측소 연동 및 Fallback 기상 정보 세팅)
            if (STATION_MAP.containsKey(country)) {
                Map<String, Object> station = STATION_MAP.get(country);
                Map<String, Object> weatherMap = new HashMap<>();
                weatherMap.put("station_id", station.get("id"));
                weatherMap.put("station_name", station.get("name"));
                weatherMap.put("temp", 22.5); // 실시간 관측 스크랩 스텁 (이후 기상 위기 발주 모드 연동)
                weatherMap.put("humidity", 60.0);
                weatherMap.put("precipitation", 0.0);
                weatherMap.put("weather_desc", "Clear sky");
                countryData.put("weather", weatherMap);
            }

            // 2. 매크로 데이터 수집 (FastAPI 및 FRED/yfinance 연동)
            Map<String, Object> macro = batchAnalysisProxyService.getMacroVector(country);
            countryData.put("macro", macro);

            // 3. GDELT 공급망 리스크 수집
            Map<String, Object> gdelt = new HashMap<>();
            gdelt.put("risk_tone", -1.25); // 정상 리스크 지수
            gdelt.put("risk_level", "NORMAL");
            countryData.put("gdelt", gdelt);

            // 4. Google Trends 시그널 수집
            Map<String, Object> trends = new HashMap<>();
            trends.put("supply_chain_trend", 45);
            trends.put("semiconductor_shortage", 12);
            countryData.put("trends", trends);

            // 타임스탬프 갱신
            countryData.put("timestamp", LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")));

            try {
                String jsonState = objectMapper.writeValueAsString(countryData);

                LkvStateId id = new LkvStateId(companyId, country);
                LkvState stateEntity = lkvStateRepository.findById(id).orElse(new LkvState());
                stateEntity.setId(id);
                stateEntity.setStateData(jsonState);
                stateEntity.setUpdatedAt(LocalDateTime.now());
                lkvStateRepository.save(stateEntity);

                log.info("SCM variables for {} saved successfully to lkv_state JSONB.", country);
            } catch (Exception e) {
                log.error("Failed to serialize LKV state for {}", country, e);
            }
        }
    }
}
